net/sockstats/sockstats_tsgo.go

```diff
@@ -122,6 +123,9 @@ func withSockStats(ctx context.Context, label Label) context.Context {
 		if sockStats.currentInterfaceCellular.Load() {
 			sockStats.rxBytesCellularMetric.Add(int64(n))
 			counters.rxBytesCellularMetric.Add(int64(n))
+			if n > 0 {
```

Comment:
I'm somewhat concerned about this adding a per-packet mutex lock + timer creation/reset overhead. An alternative design to consider is to just store an atomic int of the time of last activity, and then to have a low-frequency timer (1 Hz) that does the state transitions based on a time comparison between that int and the current time.

@raggi / @jwhited do you all have a preference from a perf perspective?

Reply:
yeah, I had the same concern.  Fortunately, there's very little in (the current) setRadioState that would take any time.  I did have an atomic int in a previous iteration, though without the timer for handling state transitions... I ended up replacing it with the mutex, but I could certainly go back to that if we're particularly concerned.  This is certainly at a low enough point in the stack that we need to be really careful.

Reply:
Is it my understanding that we're doing this for iOS only right now?

Per the brief discussion we had early on that was the case, and eschewed worrying much about performance.

If we wanted to do this for all platforms then there's a good number of considerations to expand into as the impact even on x64/arm64 at >10gbps is relevant. It is also relevant at much lower rates for our 386 and arm32 builds as the implementations don't optimize on sub-architecture features today (sse for intel, neon for arm, etc).

At the highest level, and as a gross generalization, a pointer load is the cheapest of the ops here. The Add/CAS/etc start to get into cross-domain instructions that may have a much higher cost and vary by architecture and topology. A Mutex fast path is just a CAS (when uncontended).

The timers do allocate on the heap (more than 64 bytes if I eyeballed correctly), so I share the concern about their cost. The only data needed from the data path to generate this information is a single atomic write of a boottime monotonic (that is, monotonic that ignores sleep) and everything else, current state, etc could be handled by the outside thread.

So:
- Where do we want to run (platforms, users, etc)?
- How much precision do we need?

Reply:
I've updated the implementation, but to answer your questions:

 - platforms / users: this is only included in macOS, iOS, and and android builds.  iOS and android are the ones we really care about.  This particular client metric will be on for all users.
 - precision: we can be somewhat loose, since our timings for transition states are very approximate anyway.  The new implementation only uses second precision, since we're now looking at percentage of time spent in different power states over the last hour, and transition periods are measured in whole seconds.

New implementation is doing a `time.Now()` call per packet and storing that in a fixed-size array.  All other computation is handled when reporting client metrics.

Reply:
time.Now is exceptionally slow on Xen hypervisors (about 10% or so of the Amazon fleet), as it's a work-stealing hypercall (no kvm-clock style setup, and no vdso fast-path). It's best avoided in the datapath if at all possible.

Another general issue is that CLOCK_MONOTONIC particularly for these purposes in general, may not provide the desired accuracy across sleeps (it stops on some platforms). Jason was telling me recently that he's moving WireGuard over to CLOCK_BOOTTIME.

Reply:
since we're only building this for the above platforms (enforced by build tags), that shouldn't matter in this particular case should it?

Reply:
well, put another way... how bad is time.Now() on macOS, iOS, and android?

Reply:
I guess if we're rounding to second precision anyway, we could just have a one second ticker that looks for a change in the total Tx/Rx counters to detect activity.  Then we stay out of the datapath entirely.

Reply:
I assume it's ok for those more limited, but this is one of those areas where people don't look often enough, and they get broken by sandbox changes and so on. Our using Go in a performance critical setting in these platforms is rare, so we might be the only ones who _would be affected_ - this is all to say we're going to need to test.

We can look at the implementations too, for example see `nanotime` in https://tip.golang.org/src/runtime/sys_darwin_amd64.s as you can see in the amd64 case here it varies depending on libc state, so starting to look raises more questions.

I just realized I misread "This particular client metric will be on for all users." to mean more platforms, but on re-reading I don't think that's what it meant.

As this has come up before, what I've been wanting to do, though Brad expressed an initial distaste for doing so, is to have an explicit single global clock that we step for each packet that we process, and all consumers read from that clock, with one or two places where we explicitly drive it forward. This is what netstack does, for the same reasons.

Reply:
To be clear: don't let me block you, we don't have data here. It shouldn't be catastrophic, but it may contribute to reduced performance and we'll have to come back and have a look when we study performance again. I'm more strongly concerned about macOS here than the mobiles, but there are use cases such as moving videos around from peoples primary video camera where it does matter.